packages/hardhat-chai-matchers/test/fixture-projects/hardhat-project/contracts/Matchers.sol

```diff
@@ -1,13 +1,135 @@
+// SPDX-License-Identifier: UNLICENSED
+
 pragma solidity ^0.8.0;
 
 contract Matchers {
   uint x;
 
+  AnotherContract anotherContract;
+
+  struct Pair { uint a; uint b; }
+
+  error SomeCustomError();
+  error AnotherCustomError();
+  error CustomErrorWithUint(uint);
+  error CustomErrorWithUintAndString(uint, string);
+  error CustomErrorWithPair(Pair);
+
+  constructor () {
+    anotherContract = new AnotherContract();
+  }
+
   function succeeds() public {
+    x++; // just to avoid compiler warnings
+  }
+
+  function succeedsView() public view returns (uint) {
+    return x;
+  }
+
+  function revertsWith(string memory reason) public {
     x++;
+    require(false, reason);
+  }
+
+  function revertsWithView(string memory reason) public pure {
+    require(false, reason);
   }
 
   function revertsWithoutReasonString() public {
+    x++;
     require(false);
   }
+
+  function revertsWithoutReasonStringView() public pure {
+    require(false);
+  }
+
+  function panicAssert() public {
+    x++;
+    assert(false);
+  }
+
+  function panicAssertView() public {
+    x++;
+    assert(false);
+  }
+
+  function panicUnderflow(uint n) public {
+    n--;
+    x++;
+  }
+
+  function panicUnderflowView(uint n) public pure {
+    n--;
+  }
+
+  function revertWithSomeCustomError() public {
+    x++;
+    revert SomeCustomError();
+  }
+
+  function revertWithSomeCustomErrorView() public pure {
+    revert SomeCustomError();
+  }
+
+  function revertWithAnotherCustomError() public {
+    x++;
+    revert AnotherCustomError();
+  }
+
+  function revertWithAnotherCustomErrorView() public pure {
+    revert AnotherCustomError();
+  }
+
+  function revertWithAnotherContractCustomError() public {
+    x++;
+    anotherContract.revertWithYetAnotherCustomError();
+  }
+
+  function revertWithAnotherContractCustomErrorView() public view {
```

Comment:
`view` and `pure` are equivalent in terms of how `ethers` calls those functions. I just use the "right" one in each case to prevent a compiler warning from being shown.